kythe/extractors/openjdk11/extract.go

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2019 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"bufio"
+	"errors"
+	"flag"
+	"io"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"kythe.io/kythe/go/util/flagutil"
+
+	"github.com/bazelbuild/rules_go/go/tools/bazel"
+)
+
+const (
+	javaCommandVar      = "KYTHE_JAVA_COMMAND"
+	kytheRootVar        = "KYTHE_ROOT_DIRECTORY"
+	kytheVnameVar       = "KYTHE_VNAMES"
+	javaMakeVar         = "JAVA_CMD"
+	runfilesWrapperPath = "io_kythe/kythe/extractors/openjdk11/java-wrapper"
+	runfilesVnamesPath  = "io_kythe/kythe/extractors/openjdk11/vnames.json"
+)
+
+var (
+	wrapperPath  string
+	vnameRules   string
+	errorPattern = regexp.MustCompile("ERROR: extractor failure for module ([^:]*):")
+)
+
+func setupRunfiles() error {
+	if len(os.Getenv("RUNFILES_DIR")) > 0 || len(os.Getenv("RUNFILES_MANIFEST_FILE")) > 0 {
+		return nil
+	}
+	for _, base := range []string{os.Args[0] + ".runfiles", "."} {
+		if root, err := filepath.Abs(base); err == nil {
+			if _, err := os.Stat(root); err == nil {
+				os.Setenv("RUNFILES_DIR", root)
+				if _, err := os.Stat(filepath.Join(root, "MANIFEST")); err == nil {
+					os.Setenv("RUNFILES_MANIFEST_FILE", filepath.Join(root, "MANIFEST"))
+				}
+				return nil
+			}
+		}
+	}
+	return errors.New("unable to setup runfiles")
+}
+
+func defaultWrapperPath() string {
+	val, _ := bazel.Runfile(runfilesWrapperPath)
+	return val
+}
+
+func defaultVnamesPath() string {
+	val, _ := bazel.Runfile(runfilesVnamesPath)
+	return val
+}
+
+func makeDir() string {
+	if dir := flag.Arg(0); len(dir) > 0 {
+		return dir
+	}
+	return "."
+}
+
+func findJavaCommand() (string, error) {
+	cmd := exec.Command("make", "-n", "-p")
+	cmd.Dir = makeDir()
+	cmd.Stderr = os.Stderr
+	stdout, err := cmd.StdoutPipe()
```

Comment:
Unless the output of this is really enormous, I suggest just reading the whole thing:

```go
output, err := cmd.Output()
if err != nil {
  return "", err
}
const prefix = javaMakeVar + " := "
for _, line := range strings.Split(string(output), "\n") {
  if strings.HasPrefix(line, prefix) {
```

This avoids the need for careful process handling in case of error, and doesn't create any long-term memory pressure since the program exits after doing its appointed task.

Reply:
The output is not huge, but the process take a long time (even with `-n -p`).  It's not about limiting the output, but the runtime.

Reply:
> The output is not huge, but the process take a long time (even with `-n -p`). It's not about limiting the output, but the runtime.

This makes sense, but I think the code as written may not do what you want: Once you've found the line you care about, you close its input, but that's not necessarily sufficient to terminate the process unless it's reading stdin for input. I don't think `make` does, in which case the process will likely run to completion either way. Worse, if the output is sufficient to overflow the pipe, you may have a deadlock: If you call wait without draining the process's stdout, it may block on the write and not terminate. 

So if you _do_ want to exit early, I propose you will have to kill the process (see below) before you wait for it.

```go
ctx, cancel := context.WithCancel(ctx)
defer cancel()
cmd := exec.CommandContext(ctx, "…")
…
if strings.HasPrefix(blah, prefix) {
  cancel() // safe to call this multiple times
  return blah, cmd.Wait()
}
```

or words to that effect.

Reply:
It's sufficient; I close it's output (not input) and it gets a SIGPIPE when it next attempts to write.  Although, I suspect the context is more clear, Done.

Reply:
Ah, you closed its output, not its input. My mistake.

But yes, this is still better—SIGPIPE can be suppressed by the process (and often is), so it's not guaranteed to terminate. Thanks!